### ISSUE TITLE
Poll: What do you prefer? Server or lib? Or both?

### ISSUE THREAD

**Issue by prolic** [prooph/event-store#352]
# Question for the community: What would you prefer?

### A
Connect the https://github.com/prooph/event-store-client to the "original" event-store (https://eventstore.org/)

### B
Connect the https://github.com/prooph/event-store-client to a newly written event-store server implementation written in PHP

### C
Use a simple library on top of a PDO connection (like event-store v7), no server implementation.

# You can vote multiple times, just leave a :+1: on mulitple answers.

# Pros / Cons of the different solutions:

### A
Pro: The server from https://eventstore.org/ is the best solution available, has the most features and awesome performance.

Con: Most PHP devs know a PHP / MySQL combo and would have to learn how to setup the event-store as well. Sometimes you can't make decisions on the stack used (but you can install any PHP lib you want to).

### B
Pro: Way more features are possible to implement compares to a simple library on top of a PDO connection. Communication is with client / server, not directly via PDO connection.

Con: Not all features from https://eventstore.org/ are possible to implement with PHP (but most). Communication is with client / server, not directly via PDO connection (I think some may see this as a con, so I list here again).

### C
Pro: No additional server setup needed, not https://eventstore.org/ and not a newly written server implementation in PHP.

Con: The smallest feature set of all solutions. Performance of projections is not as good.

**Comment by prolic** [comment]
# Votes for A

Connect the https://github.com/prooph/event-store-client to the "original" event-store (https://eventstore.org/)

**Comment by prolic** [comment]
# Votes for B

Connect the https://github.com/prooph/event-store-client to a newly written event-store server implementation written in PHP

**Comment by prolic** [comment]
# Votes for C

Use a simple library on top of a PDO connection (like event-store v7), no server implementation.

**Comment by basz** [comment]
> Con: Not all features from https://eventstore.org/ are possible to implement with PHP (but most). 

Can you elaborate? (I haven't used event store at all)

**Comment by prolic** [comment]
@basz just as an example: Without multithreading in PHP (yes we have pthreads, but it's really buggy) you cannot process lots of streams in parallel, hence there is a huge performance difference when projection many streams.

**Comment by enumag** [comment]
In my opinion most cases that won't enable you to choose your own dev stack won't let you install custom PHP extensions either. So while event-store-client requires protobuf and/or event loop extension, there is no point in bothering with B or C in my opinion. I think we should focus strictly on A. The con you mentioned is not even valid - if a dev is willing to learn everything around Event Sourcing and CQRS, having to use a database specifically optimized for that instead of traditional MySQL/Postgres is not a big deal at all.

**Comment by fritz-gerneth** [comment]
Does it really matter if it is the real event store or a custom PDO-based one? The client should not really care about the server-side implementation as long as it follows the same protocol? 

Other aspects I'd consider are:
- How can we achieve migration of v7 to v8? I certainly don't want to replace client & server at the same time but slowly migrate either part first
- Having too many dependencies increaes entry-level for people evaluating prooph. While PHP extensions could be solved with docker I think many people are not using it for that. Requiring an eventstore instance in addition is just another thing people quit trying. Having a super simple easy to set up minimal version for evaluation is a good thing (could even be an in-memory store ..)

**Comment by basz** [comment]
> a database specifically optimized for that instead

enumag has some good points ^^

However; A prototype developer will be familiar with PHP and PDO. Perhaps event-store has methods to manipulate events directly, but I suspect it might not be able to do that (or won't allow for such horrible practice). Being able to open a PDO connection (with DBeaver for example) to inspect and manipulate event has some practical benefits IMHO.

**Comment by Ocramius** [comment]
Unless we're talking about large scale applications, I don't see a problem with using a traditional storage layer (PDO or equivalent) for events.

For 99.5% of consumers of the library, reaching half a million events will already be a miracle, while migrating to a new technology is more effort that won't pay off, even in 10 years.

**Comment by codeliner** [comment]
I voted for A and C, here is why:

Every argument is valid because everybody is looking at the problem from their personal usage scenario.

Having A as an alternative solution next to C gives you a good choice. 

**Question is**: do we need to maintain C? I'd say yes. There are alternative implementations like Broadway or EventSauce but v7 projections are a unique feature among PHP implementations. Sure, they are not as perfect as Event Store itself, but they already do a very good job.
I'm with @fritz-gerneth that we should not break v7 entirely. @sandrokeil and @martin-schilling work on an ArangoDB v7 implementation and @sandrokeil works on a v7 MongoDB version, too (mostly driven by one of our projects because we use MongoDB there and don't want to migrate). A stable prooph event-store API is very important for those alternative implementations. Would be nice if we can keep them alive. 

It's still very important for many projects that the tech stack is kept as minimal as possible. The option to use Postgres/MySQL/MongoDB/ArangoDB as a single database for storing events and read models is a huge pro. If high throughput is needed or better stream processing capabilities EventStore is clearly the better choice. But as @Ocramius already said: 

> For 99.5% of consumers of the library, reaching half a million events will already be a miracle, while migrating to a new technology is more effort that won't pay off, even in 10 years.

I'd say prooph/event-store-client API and prooph/event-store API don't need to be compatible. We've already announced that the event-sourcing package will turn into a pure blueprint. People who want to use event-store-client should be able write their own event sourcing layer (because they should know how event sourcing works). That said, the blueprint could focus on v7/v8 only.

*Regarding B: I really love the idea but I definitely know that I won't have time for it. I'm no longer able to hack all nights on OSS and I think that's true for many prooph community members, too. If one or a group of people step up and say they want to work on a server implementation, I would offer support: review, knowledge transfer, whatever ... But actively develop the server and maintain it afterwards - no chance. The other question is if PHP is the best choice for a server. Golang for example could be a better choice for the job. It's a very nice and challenging topic, but not enough time kills it for me :(*

**Comment by Ocramius** [comment]
@codeliner I should also mentioned that I'm not scared of going into the billion events scenario with a traditional RDBMS: it works fine, and most of the problems are just with userland serialisation performance, not with the DB itself.

**Comment by bicatu** [comment]
I have similar views as @codeliner.

It is like having the choice to use a more advanced solution (eventstore) if you need and can afford the extra learning curve maintenance for new (non MySQL, PHP) stack or stick with something that has perhaps a more limited choice of features but still good enough for 99% of the usages.

So you have choice and in a runway to start simple and move if/when you need it.

**Comment by codeliner** [comment]
@Ocramius yeah, I think storing and retrieving events on the write side is no problem and performs just well, but EventStore is a way better on the projection side and with stream processing. This is why we dream of a server written for postgres. Postgres (and maybe also MySQL) could provide similar capabilities, one just needs to look at [PipelineDB](https://www.pipelinedb.com/) for example. Would be nice to have something like that working within PHP. PDO is limited to the current polling mechanism of v7 projections. It works but becomes a bottleneck relatively fast. We have no competing consumer support for example.

**Comment by oqq** [comment]
Looking at current projects, i can not see any chance to move to an event-store-client implementation.
I am not even able to move to a database infrastructure other than mysql. 

So if I am asked personally, I would always vote for having PDO as event store.

**Comment by bweston92** [comment]
Regardless of performance of MySQL some tools are just better suited.

For example we have projections running on EventStore that are updated when we append to streams.

**Comment by codeliner** [comment]
@gquemener 
> I'm interested to help in that matter

Take a look at this PR: https://github.com/sandrokeil/mongodb-event-store/pull/1/files
It already includes MongoDB change stream + a mechanism to switch from polling to listening as soon as head of the stream is reached. Something similar would be required for Postgres. But you cannot use PDO for that. [amphp/postgres](https://github.com/amphp/postgres) has Notify support. This considerations took us to the idea of a server implementation. Too much non standard stuff required hat is hard to provide as a library. A server that is just used as-is without knowing the internals could solve the problem, hence option B ;)

**Comment by darrylhein** [comment]
I haven't used prooph a lot, nor ES/CQRS, but have the simple prooph components that I didn't have to learn a many new systems (or how to configure them) helped me a lot. If it had been a lot more complicated or long to get started I likely would still be in CRUD land.

For me and the projects I work on (small to small/medium), option C is the best.

**Comment by lunetics** [comment]
I think we should go for option C:

We already have a lot of people who understand proophs event-store and can help (community). 

PHP itself (and mysql / postgres) are performant nowadays for 99,5% of event storage needs. It's only more finding a good solution for the projections. Maybe the projection runner don't even has to be based on PHP, looking into go/node direction/xyz direction.

I would see it almost as 2 different projects which can co-exist, event-store-client for real crazy needs and / or people wanting to use GetEventStore and proophs event-store for "lower" specs / needs

@codeliner If there would be more de-coupling between the eventstore and the actual projection (runner), there could be different "processors" for the projections based on e.g. amphp/postgres etc.

**Comment by prolic** [comment]
Another question: Does it have to be PDO? We could rely on amp/postgres and use it in a sync environment as well. Would allow for async event store implementation and better projection performance.
Downside: PDO doesn't work here and we need to rely on ext-pgsql or pecl-pq.

**Comment by enumag** [comment]
For async stuff I'd much prefer Swoole coroutine over AMP.

**Comment by prolic** [comment]
Swoole and postgres doesn't work together (yet). But you can use any amp lib within swoole or react-php, hence you have a lot of flexibility.

**Comment by prolic** [comment]
Only postgres has the polling support which is great for projections.
However we could use mysql and query in a loop like we did before.

**Comment by codeliner** [comment]
@lunetics 
> If there would be more de-coupling between the eventstore and the actual projection (runner), there could be different "processors" for the projections based on e.g. amphp/postgres etc.

Good point!

@prolic What do you think? One could use PDO for writing and amphp/postgres on projection side. If we take the lessons learned from v7 - mainly that stream strategies are too complex and one table for all events works best - and simplify the low level API then the projection layer could be simpler as well and we could ship different variations.

**Comment by prolic** [comment]
@codeliner I just try to limit the scope a little. Of course we can ship a pdo-event-store + and async amp-event-store, but this doubles the maintenance. But an amp event-store alone can also work in traditional sync environments (but it's blocking then) and you can use it in react-php and swoole.
Maintaining PDO additionally duplicates a lot of work. Hence the question: Do we really NEED to rely on PDO or is it okay for people to install ext-pgsql or pecl-pq instead of PDO?
Note: When using amp we don't need any mysql extension, but only postgres has the polling feature.

**Comment by codeliner** [comment]
@enumag Swoole supports async Postgres, that's correct. But it does not support async PDO Postgres connection. The latter would only work for our PDO MySQL event store.

Btw. I got a hint from Swoole team that you can add coroutine support to other PHP extensions as well. You need to be a little bit crazy but here is how it works: https://translate.google.com/translate?hl=de&sl=auto&tl=en&u=https%3A%2F%2Fwiki.swoole.com%2Fwiki%2Fpage%2F983.html&sandbox=1

*The original Swoole docs contain many interesting details. Unfortunately, you need google translate to understand something.*

**Comment by fritz-gerneth** [comment]
> Maintaining PDO additionally duplicates a lot of work. Hence the question: Do we really NEED to rely on PDO or is it okay for people to install ext-pgsql or pecl-pq instead of PDO?

I think MySQL support is needed: the argument for option C in here has been to provide as wide support as possible for learning / low-volumne use cases. For anything with high throughput / best hardware usage either option A or a custom projection implenentation in any non-php language is the way to go. This does not necessarily need to be backed up by PDO though. 
As @enumag already was asking, if different DB platforms could be supported by AMP this would be the best solution for me (and just hint to the downsides of individual adapters). 

Remembering the discussion about Swoole recently, I really would like to use it but currently see no option to do though. Documentation language and availability of information is the primary concern here (which has been demonstrated in this discussion already). Also this too follows the same direction that option C should be beginner friendly. Asking them to install Swoole first isn't. 

PG is out of questions for us at all. We are using MySQL quite fine. It is the platform we have most experience with and much of our non-prooph code is dependent on. Introducing new elements to our stack for now real benefit is not going to happen (esp a new DB platform, introducing a whole new set of operations issues we have no experience with). E.g. adding more hardware for scaling is cheaper for us for now, should be outgrow MySQL we probably would move to ES right away. 

My point here: in many cases chosing the DB platform is not a choice. Locking support to one specific platform drives people away / makes them stay away (I probably would not have started to use prooph if it were only supported PG). 

I'm in strong favor of a MVP option C with as wide support as possible and as little requirements as possible. If more advanved features can be plugged in / added side-ways this is great, but should not be done on the expanse of adding requirements (so no specific DB platform or php extension requirements, composer deps are fine).

**Comment by prolic** [comment]
Just to be clear: amp + mysql will NOT NEED any PHP extension, not even PDO or mysqli. This can be used everywhere, in traditional sync application, in react-php and in swoole apps.

amp + postgres will NOT work with PDO, but with ext-pgsql or pecl-pq. The postgres solution would have polling feature and be much more performant.

There is no plan to drop mysql support. Mysql lacks the polling feature and there is some code that needs to different (query in loop vs polling) but that's okay.

What I do not want to do (because lack of time): Provide a PDO implementation + a distinct async implementation, because that's double the amount of code to write and maintain.

**Comment by fjogeleit** [comment]
Maybe a help for improving projection performance are optional plugins. I wrote a small plugin for my project to trigger projections over the redis pub/sub feature. RabbitMQ or other queue based technology are also possible. So its not related to the database.

**Comment by MassiAtHG** [comment]
Now that Mongo 4 supports ACID transactions, can it become a valuable persistency layer?

**Comment by prolic** [comment]
To me event-store is about the event store, not about which database we
have in background. I don't care.

**Comment by sandrokeil** [comment]
@MassiAtHG New MongoDB Event Store is [in progress](https://github.com/sandrokeil/mongodb-event-store/pull/1), but it looks like there are some transaction issues in current MongoDB version. So maybe we have to wait for MongoDB 4.2. You will see it in the Event Store Benchmark.

**Comment by prolic** [comment]
This poll was very telling. Event-store-client will have final release about end of this year, docs are currently missing. There is no real demand for an eventstore server implementation written in PHP, and it takes incredible amount of work to do it, so there will be none in the foreseeable future. Eventstore v8 development will start once event-store-client final release is done.
